File: leetcode/1203.cc
#include <vector>


using namespace std;


enum Color {
    kWhite,
    kGray,
    kBlack,
};


class Solution {
  public:
    bool TopologicalSort(int cur, vector<vector<int>>& adjs,
                         vector<int>* visited_ptr, vector<int>* result_ptr) {
        auto& visited = *visited_ptr;
        visited[cur] = kGray;

        for (auto& adj : adjs[n]) {
            if (visited[adj] == kGray)
                return false;
            if (visited[adj] == kBlack)
                continue;

            if (!TopologicalSort(adj, adjs, visited_ptr, result_ptr))
                return false;
        }
        result_ptr->emplace_back(cur);
        return true;
    }

    vector<int> sortItems(int n, int m, vector<int>& group, vector<vector<int>>& before) {
        int g_cnt = m;
        for (int i = 0; i < n; ++i) {
            if (group[i] != -1)
                continue;
            group[i] = g_cnt;
            ++g_cnt;
        }

        vector<int> visited(n, kWhite);
        vector<int> ordred;
        for (int i = 0; i < n; ++i) {
            if (visited[i] == kWhite)
                TopologicalSort(i, before, &visited, &ordered);
        }
    }
};

File: leetcode/1203.py
import enum


class Color(enum.Enum):
    WHITE = 0
    GRAY = 1
    BLACK = 2


class Solution:
    def topological_sort(self, adjs):
        n = len(adjs)
        visited = [Color.WHITE] * n
        order = []

        def dfs(u):
            visited[u] = Color.GRAY
            for v in adjs[u]:
                if visited[v] == Color.BLACK:
                    continue
                elif visited[v] == Color.GRAY:
                    return False

                if not dfs(v):
                    return False
            visited[u] = Color.BLACK
            order.append(u)

            return True

        for s in range(n):
            if visited[s] == Color.BLACK:
                continue
            assert visited[s] == Color.WHITE

            if not dfs(s):
                return None
        return order

    def sortItems(self, n: int, m: int, group: List[int], beforeItems: List[List[int]]) -> List[int]:
        for i in range(n):
            if group[i] == -1:
                group[i] = m
                m += 1

        item_order = self.topological_sort(beforeItems)
        if item_order is None:
            return []

        item_loc = [-1] * n
        for i, v in enumerate(item_order):
            item_loc[v] = i

        before_groups = [[] for _ in range(m)]
        for i, before in enumerate(beforeItems):
            gi = group[i]
            for b in before:
                gb = group[b]
                if gi != gb:
                    before_groups[gi].append(gb)
        group_order = self.topological_sort(before_groups)
        if group_order is None:
            return []

        group_loc = [-1] * m
        for i, v in enumerate(group_order):
            group_loc[v] = i

        return sorted(range(n), key=lambda x: (group_loc[group[x]], item_loc[x]))
